lib/api/objectGet.js

```diff
@@ -20,30 +23,57 @@ function objectGet(authInfo, request, log, callback) {
     log.debug('processing request', { method: 'objectGet' });
     const bucketName = request.bucketName;
     const objectKey = request.objectKey;
+
+    const decryptVidResult = decryptVersionId(request.query);
+    if (decryptVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decryptVidResult,
+        });
+        return callback(decryptVidResult);
+    }
+    const versionId = decryptVidResult;
+
     const mdValParams = {
         authInfo,
         bucketName,
         objectKey,
+        versionId,
         requestType: 'objectGet',
-        log,
     };
 
-    services.metadataValidateAuthorization(mdValParams, (err, bucket,
+    return metadataValidateBucketAndObj(mdValParams, log, (err, bucket,
         objMD) => {
         const corsHeaders = collectCorsHeaders(request.headers.origin,
             request.method, bucket);
         if (err) {
-            log.debug('error processing request', { error: err });
             return callback(err, null, corsHeaders);
         }
         if (!objMD) {
-            return callback(errors.NoSuchKey, null, corsHeaders);
+            const err = versionId ? errors.NoSuchVersion : errors.NoSuchKey;
+            return callback(err, null, corsHeaders);
+        }
+        const verCfg = bucket.getVersioningConfiguration();
+        if (objMD.isDeleteMarker) {
+            const responseMetaHeaders = Object.assign({},
+                { 'x-amz-delete-marker': true }, corsHeaders);
+            if (!versionId) {
```

Comment:
add comment: if general get request without versionId and master version is a delete marker, return NoSuchKey